Fixed #32948 -- Optimized django.utils.tree.Node and its subclasses.

ngnpope opened the pull request:
ticket-32948

An alternative to #14673 with extra simplifications and optimisations.

kezabelle commented on django/db/models/sql/where.py:
```diff
@@ -58,8 +58,8 @@ def split_having(self, negated=False):
                 where_parts.append(c)
-        having_node = self.__class__(having_parts, self.connector, self.negated) if having_parts else None
-        where_node = self.__class__(where_parts, self.connector, self.negated) if where_parts else None
+        having_node = self.create(having_parts, self.connector, self.negated) if having_parts else None
+        where_node = self.create(where_parts, self.connector, self.negated) if where_parts else None
```
So I'm going through your changes and figuring out which bits to pull back in as per Mariusz' request. This one confuses me, as I'm intending to pull 2399d5da1d2f842be45fee17dacb52e68592737f which though I've not benchmarked it, probably makes sense. It sounds correct, anyway:

> As `WhereNode` doesn't override `__init__()` we can just use `self.__class__()` directly which is faster. As an aside, we also do not need to pass in `children=[]`. That is handled by `Node.__init__()`

But here you've gone back to using `create`/`_new_instance`. Is that because the `__class__` can end up as a `Q` rather than as I'd guess, a `WhereNode`?

ngnpope replied:
Heya. Just seen this. Will reply later this evening. 🙂

ngnpope replied:
Hey @kezabelle. So I mentioned about bringing `._new_instance()` back in https://github.com/django/django/pull/14673#discussion_r673884166 but I wanted to "bless" it more by renaming it to `.create()` (while still keeping it private).

This isn't really going to be ready for the feature freeze, so let's not rush it. I'm hoping to give it some attention over the next couple of weeks. My thoughts for this were:

- Pick "Remove obsolete docstring from `WhereNode.clone()`."
- Bring in your changes from #14709.
- Rename `Node._new_instance()` to `Node.create()`.
- Make use of `.create()` to replace calls using `type(self)(...)`, `self.__class__(...)`, `self.__class__.create(...)`, and `Node(...)`.
- I also figured we could use it to follow up on #14699 and switch (again) to use `.create()` to avoid `*`-unpacking and the `super().__init__()` call in `Q.__init__()`. It'll make most of the construction of these `Node` (sub)classes consistent.
- And then, as per the end of the description on ticket-32970 we implement `__copy__` and `__deepcopy__` with `.copy()` and `.clone()` as aliases to `__copy__`. We put this on `Node` as all subclasses are pretty much doing the same thing. (The aliases are just for backward compatibility, for now at least).
- And then there is just the tidying of the messy if-statements, etc.

Just to note, because of the way `._new_instance()/.create()` is implemented, I expect that we just end up with:

```python
class Node:
    def __init__(self, children=None, connector=None, negated=False):
        self.children = children[:] if children else []
        self.connector = connector or self.default
        self.negated = negated

    @classmethod
    def create(cls, children=None, connector=None, negated=False):
        obj = Node(children, connector, negated)
        obj.__class__ = cls
        return obj

    def __copy__(self):
        return self.create(children=self.children, connector=self.connector, negated=self.negated)

    copy = __copy__
    clone = __copy__

    def __deepcopy__(self, memodict):
        obj = self.create(connector=self.connector, negated=self.negated)
        obj.children = copy.deepcopy(self.children, memodict)
        return obj
```

kezabelle replied:
I'm glad I asked. I sincerely appreciate the amount of thought you've put into it, compared to my piecemeal scattergun approach across multiple PRs :)

ngnpope replied:
@kezabelle So I've quickly hacked together what I think this ought to look like. The intent was to try and encapsulate the funny behaviour of setting `__class__` inside `Node.create()` and then use that everywhere.

(After playing with this, a further thing to do that would be nice is to unify `.copy()` and `.clone()` if possible.)

ngnpope commented on django/utils/tree.py:
```diff
@@ -32,7 +32,7 @@ def create(cls, children=None, connector=None, negated=False):
         Node.__init__().
         """
-        obj = Node(children, connector, negated)
+        obj = Node(children, connector or cls.default, negated)
```
This was necessary to fix an issue that prevented `.create()` (previously `._new_instance()`) from being used by `Q()`. This is because `Node.__init__()` is called which falls back to `Node.default`, not `Q.default` as the default value for `connector`.

ngnpope commented on django/utils/tree.py:
```diff
@@ -47,9 +43,14 @@ def __str__(self):
     def __repr__(self):
         return "<%s: %s>" % (self.__class__.__name__, self)
 
+    def __copy__(self):
+        return self.create(self.children, self.connector, self.negated)
+
+    copy = __copy__
+    clone = __copy__
```
Alas, we get a single test failure, confusingly in a test for `__deepcopy__`:

```python
======================================================================
FAIL: test_deepcopy (utils_tests.test_tree.NodeTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tests/django/tests/utils_tests/test_tree.py", line 76, in test_deepcopy
    self.assertIs(self.node1.children, node4.children)
AssertionError: [('a', 1), ('b', 2)] is not [('a', 1), ('b', 2)]
```

This is actually a check for shallow copy, but our `__copy__` here is using `.create()` which calls `children[:]`. This is fixed by rolling back to the following:

```diff
diff --git a/django/utils/tree.py b/django/utils/tree.py
index 67b47f027c..8bc01d85fb 100644
--- a/django/utils/tree.py
+++ b/django/utils/tree.py
@@ -44,10 +44,14 @@ def __repr__(self):
         return "<%s: %s>" % (self.__class__.__name__, self)
 
     def __copy__(self):
-        return self.create(self.children, self.connector, self.negated)
+        obj = self.create()
+        obj.__dict__ = self.__dict__.copy()
+        return obj
 
     copy = __copy__
-    clone = __copy__
+
+    def clone(self):
+        return self.create(self.children, self.connector, self.negated)
 
     def __deepcopy__(self, memodict):
         obj = self.create(connector=self.connector, negated=self.negated)
```

So we have two options:

1. Keep it as is and don't worry that the copy is not as shallow... 😬 
2. Apply this patch.

If we go for option 2 we probably ought to check other uses of `.copy()` to decide whether we need to be using `.clone()` instead, notably in `.add()`, or we need to determine whether `.clone()` is really meant to be a `__deepcopy__` alias.

I don't really understand in a structure like this why we don't just have shallow and deep copying, but have this funny extra thing to clone one level down. Surely if we nest these objects then we want to deep copy to avoid any issues with changes deeper inside the tree?

@kezabelle Would you be able to run some performance checks over this? I'm not sure it'll necessarily be as good as I first had, but it shouldn't be worse than the original. I think that the tidy up to have a consistent approach here is most important, however, as it was getting crufty.

kezabelle replied:
Will do, marking it as unread to remind me.

ngnpope replied:
Ok, so I've reverted `__copy__()` to something similar in layout to `__deepcopy__()`:

```python
    def __copy__(self):
        obj = self.create(connector=self.connector, negated=self.negated)
        obj.children = self.children  # Don't [:] as .__init__() via .create() does.
        return obj
```

ngnpope commented:
I've updated this following #14916. I think it should be ready now.

ngnpope commented:
I ran some basic checks on the performance - seems pretty good!

```bash
$ sudo pyperf system tune

# Before:

$ pyperf timeit --setup 'from django.db.models import Q' --setup 'a = Q(a=1, b=2, c=3)' --setup 'b = Q(d=4, e=5, f=6)' 'a & b'
.....................
Mean +- std dev: 2.46 us +- 0.06 us

$ pyperf timeit --setup 'from django.db.models import Q' --setup 'a = Q(a=1, b=2, c=3)' --setup 'b = Q(d=4, e=5, f=6)' 'a | b'
.....................
Mean +- std dev: 2.79 us +- 0.04 us

$ pyperf timeit --setup 'from django.db.models import Q' --setup 'a = Q(a=1, b=2, c=3)' --setup 'b = Q(d=4, e=5, f=6)' '~a'
.....................
Mean +- std dev: 1.65 us +- 0.02 us

# After:

$ pyperf timeit --setup 'from django.db.models import Q' --setup 'a = Q(a=1, b=2, c=3)' --setup 'b = Q(d=4, e=5, f=6)' 'a & b'
.....................
Mean +- std dev: 1.96 us +- 0.04 us

$ pyperf timeit --setup 'from django.db.models import Q' --setup 'a = Q(a=1, b=2, c=3)' --setup 'b = Q(d=4, e=5, f=6)' 'a | b'
.....................
Mean +- std dev: 2.29 us +- 0.04 us

$ pyperf timeit --setup 'from django.db.models import Q' --setup 'a = Q(a=1, b=2, c=3)' --setup 'b = Q(d=4, e=5, f=6)' '~a'
.....................
Mean +- std dev: 1.00 us +- 0.02 us

$ sudo pyperf system reset
```

kezabelle commented:
I still need to come back and sit down and look at the _implementation_ again, but I have finally sat down to compare numbers vs the baseline & the original solution I was proposing (see timeit in ticket), and it's favourable:

Baseline:
```
In [5]: %timeit ~(full | full2)
5.04 µs ± 78 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [6]: %timeit ~(full & full2)
4.83 µs ± 108 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [10]: %timeit reduce(or_, values)
12.1 µs ± 141 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

This branch:
```
In [6]: %timeit ~(full | full2)
3.45 µs ± 33.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [7]: %timeit ~(full & full2)
3.18 µs ± 59.3 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [9]: %timeit reduce(or_, values)
9.64 µs ± 132 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
So I definitely think we can run with it.

ngnpope commented:
> I still need to come back and sit down and look at the _implementation_ again, ...

Yes - it'll be good to check that I've made reasonable decisions. I have tried to explain in the commit message for each commit why I did things, so that should help with the reviewing.

> ...but I have finally sat down to compare numbers vs the baseline & the original solution I was proposing (see timeit in ticket), and it's favourable:

That's good to know. \o/

Thanks @kezabelle.

kezabelle commented:
So, I've sat down and attempted to work through it, trying to remember what I'd originally planned and where you've ended up subsequently. I _think_ it may fine, and probably tidier than the place I ended up; I can't easily reason about the `if` branch changes made in `_combine` because I'm less familiar with what's considered a combinable, and conditional etc. But I trust that it's correct.

Given the unfortunate situation with the attempt to optimise WhereNode cloning (See ticket [33159](https://code.djangoproject.com/ticket/33159), thank goodness for beta testers.) I'm a little cautious of this line precisely because it doesn't copy at all (I think):
```
obj.children = self.children # Don't [:] as .__init__() via .create() does.
```
(caveat: I haven't checked whether you or I introduced it, I may be picking holes in my own proposal for all I now remember.)
I think it'd be good to have that well tested, if it's not already, as it's only a reference right? e.g.: because `add` may `extend` or `append` to `children` we want to be super sure that a later operations on a Q doesn't backfill changes to a previously declared Q which looks untouched. Logic notwithstanding, here's the kind of thing.
```
a = Q(a=1) # this isn't directly modified and looks innocuous.
b = Q(b=2)
c = a & Q(c=3). # a doesn't get mutated here, or by any other operations that can lead to combine/add happening?
```

felixxm commented:
@ngnpope Do you have time to keep working on this?

ngnpope commented:
I'll rebase and have another look next week. I think it was basically there, just needed to check whether existing tests covered the expected behaviour well enough.

ngnpope commented:
Ok. I've rebased this. To satisfy the concerns around issues with mutating existing objects when copying I added/enhanced some tests to properly capture the existing behaviour to ensure that hasn't changed - basically ensuring that my bullet points in my previous [comment](https://github.com/django/django/pull/14677#issuecomment-946618000) hold true.

Should be good to land.

felixxm commented:
@kezabelle Do you want to take a look?

felixxm commented:
I merged 769d7cce4aedfcbba59f1b68577225d07701c206, 4f8bc75bc310a6a206d2d1b9550d3df313eb63eb, and cc52e02c96e9e26d59c6512f18516657e353cbe1, and rebased from the `main` branch.

felixxm commented:
@ngnpope @kezabelle Thanks both :rocket: I reorganized commits.

kezabelle commented:
It's been a fair while since I/we looked at this (thank you so much for coming back to it), so it'd take me a bit of time to get back into any of the nuance we've previously covered, but at first blush this looks good to me.
